tt_metal/hw/firmware/src/brisc.cc

```diff
@@ -433,10 +443,13 @@ int main() {
             WAYPOINT("R");
             if (enables & DISPATCH_CLASS_MASK_TENSIX_ENABLE_DM0) {
                 setup_cb_read_write_interfaces(cb_l1_base, num_cbs_to_early_init, launch_msg_address->kernel_config.max_cb_index, true, true, false);
-                kernel_init();
+                int index = static_cast<std::underlying_type<TensixProcessorTypes>::type>(TensixProcessorTypes::BRISC);
+                void (*kernel_address)(uint32_t) = (void (*)(uint32_t))
+                    (kernel_config_base + launch_msg_address->kernel_config.kernel_text_offset[index]);
+                (*kernel_address)((uint32_t)kernel_address);
```

Comment:
why does the kernel need to be passed its own address? If it needs it it can compute it itself.

Reply:
The kernel address is passed to the firmware.  The firmware was linked separately and doesn't know the kernel addresses as these are calculated at runtime based on the ring buffer location and offset.

Reply:
oh, wait, realized what you meant - the argument.  it doesn't "need" it, but the value will be sitting in a register and so passing it is free while calculating it is not

Reply:
The first instruction is the start of the text section.  The linker script arranges that that's where the entry point is, and the elfloader detects the text section that way -- so if the invariant is violated, the loader barfs. The firmware doesn't need to know symbols in the kernel -- just the segments.

Why does the startup code need its own address though?

Reply:
started backing out this change when I realized the other motivation for this pattern:
ncrisck can’t know the address, it has to be passed in since the address it would calculate is in IRAM not in the ring buffer, so it can’t find its locals that need to be copied.  so, everything else matches that pattern